agent.go

```diff
@@ -319,9 +321,25 @@ func (a *Agent) onSelectedCandidatePairChange(p *candidatePair) {
 
 func (a *Agent) startOnConnectionStateChangeRoutine() {
 	go func() {
-		for s := range a.chanState {
-			if hdlr, ok := a.onConnectionStateChangeHdlr.Load().(func(ConnectionState)); ok {
-				hdlr(s)
+		for {
+			select {
+			case s, isOpen := <-a.chanState:
+				if !isOpen {
+					return
+				}
+
+				if hdlr, ok := a.onConnectionStateChangeHdlr.Load().(func(ConnectionState)); ok {
+					hdlr(s)
+				}
+
+			case c, isOpen := <-a.chanCandidate:
+				if !isOpen {
+					return
```

Comment:
so, maybe we need to add
```go
for state := range a.chanState {
}
```
to avoid missing another chan's buffer?

Reply:
Sorry I don't fully understand. You can push directly to the branch if you want! Or paste the snippet here, happy to drop it in :)

Reply:
ok, let me write some code

Reply:
It's safe if both `chanState` and `chanCandidate` are not buffered.
So, https://github.com/pion/ice/pull/224#discussion_r449405079 made this part safe.

if buffered, it may drop one.
https://play.golang.org/p/QDr0R4x0PNo

Reply:
I mean:
```suggestion
					for c := range a.chanCandidate {
						if onCandidateHdlr, ok := a.onCandidateHdlr.Load().(func(Candidate)); ok {
							onCandidateHdlr(c)
						}
					}
					return
```
to flush data in the chan.

Reply:
let me pick code from https://github.com/pion/ice/pull/225